src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,1052 @@
+        self.dropout = nn.Dropout(p=pooler_dropout)
+        self.out_proj = nn.Linear(inner_dim, num_classes)
+
+    def forward(self, x):
+        x = self.dropout(x)
+        x = self.dense(x)
+        x = torch.tanh(x)
+        x = self.dropout(x)
+        x = self.out_proj(x)
+        return x
+
+
+class LearnedPositionalEmbedding(nn.Embedding):
+    """
+    This module learns positional embeddings up to a fixed maximum size.
+    Padding ids are ignored by either offsetting based on padding_idx
+    or by setting padding_idx to None and ensuring that the appropriate
+    position ids are passed to the forward function.
+    """
+
+    def __init__(
+        self, num_embeddings: int, embedding_dim: int, padding_idx: int,
+    ):
+        # if padding_idx is specified then offset the embedding ids by
+        # this index and adjust num_embeddings appropriately
+        assert padding_idx is not None
+        num_embeddings += padding_idx + 1  # WHY?
+        super().__init__(num_embeddings, embedding_dim, padding_idx=padding_idx)
+
+    def forward(self, input):
+        """Input is expected to be of size [bsz x seqlen]."""
+        positions = create_position_ids_from_input_ids(input, self.padding_idx)
+        return super().forward(positions)
+
+
+def LayerNorm(normalized_shape, eps=1e-5, elementwise_affine=True):
+    if torch.cuda.is_available():
+        try:
+            from apex.normalization import FusedLayerNorm
+
+            return FusedLayerNorm(normalized_shape, eps, elementwise_affine)
+        except ImportError:
+            pass
+    return torch.nn.LayerNorm(normalized_shape, eps, elementwise_affine)
+
+
+def fill_with_neg_inf(t):
+    """FP16-compatible function that fills a input_ids with -inf."""
+    return t.float().fill_(float("-inf")).type_as(t)
+
+
+def _filter_out_falsey_values(tup) -> Tuple:
+    """Remove entries that are None or [] from an iterable."""
+    return tuple(x for x in tup if isinstance(x, torch.Tensor) or x)
+
+
+# Public API
+
+
+@add_start_docstrings(
+    "The bare BART Model outputting raw hidden-states without any specific head on top.",
+    BART_START_DOCSTRING,
+    BART_INPUTS_DOCSTRING,
+)
+class BartModel(PretrainedBartModel):
+    """"""
+
+    def __init__(self, config: BartConfig):
+        super().__init__(config)
+        self.output_attentions = config.output_attentions
+        self.output_hidden_states = config.output_hidden_states
+
+        padding_idx, vocab_size = config.pad_token_id, config.vocab_size
+        self.shared = nn.Embedding(vocab_size, config.d_model, padding_idx)
+
+        self.encoder = BartEncoder(config, self.shared)
+        self.decoder = BartDecoder(config, self.shared)
+
+        self.init_weights()
+
+    def get_input_embeddings(self):
+        return self.shared
+
+    def set_input_embeddings(self, value):
+        self.shared = value
+
+    def get_output_embeddings(self):
+        return _make_linear_from_emb(self.shared)
+
+    def forward(
```

Comment:
We now link to the inputs in the forward method, cf. [BERT file](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_bert.py#L635)